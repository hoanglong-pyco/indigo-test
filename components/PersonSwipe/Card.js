import React from "react";
import styled from "styled-components/native";

export function Card({ children, ...props }) {
  return (
    <Wrapper {...props}>
      <BackgroudTop />
      <Container>{children}</Container>
    </Wrapper>
  );
}

const styles = {
  rarius: 8,
  width: 280,
  borderColor: "#ddd"
};

const Wrapper = styled.View`
  width: ${styles.width};
  height: 300;
  background-color: #fff;
  border-width: 1;
  border-color: ${styles.borderColor};
  border-radius: ${styles.rarius};
`;
const BackgroudTop = styled.View`
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  height: 105;
  background-color: #f9f9f9;
  border-bottom-width: 1;
  border-color: ${styles.borderColor};
  border-top-left-radius: ${styles.rarius - 1};
  border-top-right-radius: ${styles.rarius - 1};
`;

const Container = styled.View`
  width: ${styles.width};
  flex: 1;
  height: 400;
  align-items: center;
`;
