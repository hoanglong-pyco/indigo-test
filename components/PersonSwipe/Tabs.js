import React, { memo, useState, useEffect } from "react";

import styled from "styled-components/native";

import { View, TouchableOpacity } from "react-native";

import { Ionicons } from "@expo/vector-icons";

export function Tabs({ items, defaultSelect = 0 }) {
  const [selected, setSelected] = useState(defaultSelect);
  const [_, SelectedPannel] = items[selected];
  return (
    <Wrapper>
      <TabContent>
        <SelectedPannel />
      </TabContent>
      <Tabbar>
        {items.map(([icon], index) => (
          <TabIcon
            key={icon}
            active={index === selected}
            icon={icon}
            onPress={() => setSelected(index)}
          />
        ))}
      </Tabbar>
    </Wrapper>
  );
}
const Wrapper = styled.View`
  height: 110;
  align-items: center;
`;
const Tabbar = styled.View`
  height: 40;
  flex-direction: row;
`;
const TabContent = styled.View`
  height: 90;
`;

const TabIcon = ({ active, icon, onPress }) => {
  return (
    <TouchableOpacity style={{ padding: 5 }} onPress={onPress}>
      <Ionicons name={icon} size={32} color={active ? "#87ba4f" : "#d9d9d9"} />
    </TouchableOpacity>
  );
};
