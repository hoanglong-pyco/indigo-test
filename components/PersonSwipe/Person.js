import React, { memo } from "react";
import { View } from "react-native";

import styled from "styled-components/native";

import { Card } from "./Card";
import { Tabs } from "./Tabs";

export const Person = memo(
  ({ picture, location: { street, city }, phone, name, email, gender }) => {
    return (
      <Card>
        <CoverWrap>
          <CoverImage
            source={{ uri: "https://picsum.photos/id/119/240/240" }}
          />
        </CoverWrap>
        <Tabs
          defaultSelect={2}
          items={[
            [
              "ios-person",
              simplePanel({
                label: [
                  `${name.title}: ${name.first} ${name.last}`,
                  `Mail: ${email}`,
                  `Gender: ${gender.toUpperCase()}`
                ].join("\n")
              })
            ],
            ["ios-list-box", simplePanel({ label: "Commingsoon" })],
            [
              "ios-map",
              simplePanel({
                label: "My address is",
                value: [street, city].join(", ")
              })
            ],
            ["ios-call", simplePanel({ label: "My phone is", value: phone })],
            ["ios-lock", simplePanel({ label: "Commingsoon" })]
          ]}
        />
      </Card>
    );
  }
);

const CoverWrap = styled.View`
  width: 130;
  height: 130;
  background-color: #fff;
  border-color: #ddd;
  border-width: 1;
  border-radius: 65;
  margin-top: 15;
  margin-bottom: 15;
  align-items: center;
  justify-content: center;
`;
const CoverImage = styled.Image`
  width: 120;
  height: 120;
  background-color: #fafafa;
  border-radius: 60;
`;

const simplePanel = ({ label, value }) => () => {
  return (
    <View>
      <SimplePanelLabel>{label}</SimplePanelLabel>
      <SimplePanelValue>{value}</SimplePanelValue>
    </View>
  );
};

const SimplePanelLabel = styled.Text`
  font-size: 14;
  color: #9b9b9b;
  text-align: center;
`;
const SimplePanelValue = styled.Text`
  font-size: 17;
  color: #2b2e31;
  text-align: center;
  padding-left: 10;
  padding-right: 10;
`;
