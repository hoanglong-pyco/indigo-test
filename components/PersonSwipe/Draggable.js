import React, { useState, useEffect } from "react";
import { PanResponder, Animated, Dimensions } from "react-native";

import { Person } from "./Person";

const opacityWithDx = () => {
  const width75percent = Math.round(Dimensions.get("window").width) * 0.75;
  return dx => 1 - Math.min(Math.abs(dx / width75percent), 1)
};

const animateWithProps = (props, duration) => {
  return (toProps = {}, onEnd = Boolean) => {
    Object.entries(toProps).forEach(([name, toValue]) => {
      Animated.timing(props[name], {
        duration,
        toValue,
      }).start();
    });
    setTimeout(onEnd, duration);
  };
};

export function Draggabe({
  children,
  onMove = Boolean,
  onAnimate = Boolean,
  onEnd = Boolean
}) {
  const [css] = useState({
    opacity: new Animated.Value(1),
    position: new Animated.ValueXY(0, 0),
    rotate: new Animated.Value(0)
  });
  const getOpacityDx = opacityWithDx();
  const animate = animateWithProps(css, 300);
  const { opacity, position, rotate } = css;

  const pan = PanResponder.create({
    onMoveShouldSetPanResponder: () => true,
    onMoveShouldSetPanResponderCapture: () => true,
    onPanResponderMove(e, gestureState) {
      const { dx, dy } = gestureState;
      position.setValue({ x: dx, y: dy });
      rotate.setValue(dx / 3000);
      opacity.setValue(getOpacityDx(dx));
      onMove(e, gestureState);
    },
    onPanResponderRelease(e, gestureState) {
      let { vx, vy, dx, dy } = gestureState;
      let toRotate = 0,
        toOpacity = 1,
        duration = 300;

      dx += vx * duration;

      if (getOpacityDx(dx) > 0.5) {
        dx = 0;
        dy = 0;
      } else {
        dy += vy * duration;
        toRotate = dx / 3000;
        toOpacity = 0;
      }
      const endGestureState = Object.assign({}, gestureState, { dx, dy });
      animate(
        {
          opacity: toOpacity,
          position: { x: dx, y: dy },
          rotate: toRotate
        },
        () => onEnd(e, endGestureState)
      );
      onAnimate(e, endGestureState);
    }
  });

  return (
    <Animated.View
      style={{
        opacity: css.opacity,
        transform: [
          { translateX: css.position.x },
          { translateY: css.position.y },
          {
            rotateZ: css.rotate.interpolate({
              inputRange: [0, 1],
              outputRange: ["0deg", "360deg"]
            })
          }
        ]
      }}
      Î
      {...pan.panHandlers}
    >
      {children}
    </Animated.View>
  );
}
