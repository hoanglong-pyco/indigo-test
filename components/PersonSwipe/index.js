import React, { memo, useState, useEffect } from "react";
import { ActivityIndicator, View, Animated, Easing } from "react-native";

import { Draggabe } from "./Draggable";
import { Person } from "./Person";

export * from './Person';

export default memo(({
  onSwipeLeft = Boolean,
  onSwipeRight = Boolean,
  onDrop = Boolean,
  person
}) => {
  const [fadeAnim] = useState(new Animated.Value(0));
  useEffect(() => {
    person &&
      Animated.timing(fadeAnim, {
        toValue: 1,
        // easing: Easing.back(),
        duration: 300,
        useNativeDriver: true
      }).start();
  }, [person]);

  if (!person) return <ActivityIndicator />;
  return (
    <Draggabe
      onAnimate={(e, { dx }) => {
        if (dx > 0) {
          onSwipeRight(person);
        } else if (dx < 0) {
          onSwipeLeft(person);
        }
      }}
      onEnd={(e, {dx}) => dx === 0 || onDrop(person)}
    >
      <Animated.View style={{ opacity: fadeAnim }}>
        <Person {...person} />
      </Animated.View>
    </Draggabe>
  );
}, Boolean )
