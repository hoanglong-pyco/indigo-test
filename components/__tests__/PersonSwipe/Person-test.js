import React from "react";
import renderer from "react-test-renderer";
import "jest-styled-components";

import { Person } from "../../PersonSwipe";
const user = {
  gender: "female",
  name: { title: "ms", first: "anna", last: "wheeler" },
  location: {
    street: "7403 long rapids rd",
    city: "cedar rapids",
    state: "south dakota",
    zip: "99516"
  },
  email: "anna.wheeler18@example.com",
  username: "goldenfish705",
  password: "toriamos",
  salt: "0k4z47Gb",
  md5: "e848d0ad805643a39458453a4f128826",
  sha1: "1351838306c05a89c6ad61ad91691691bac922fb",
  sha256: "d18d420793e57f14da3ebeccf4053b3305e96c8c48f1cb2cc0ca359bc5abcf08",
  registered: "1123380414",
  dob: "178768751",
  phone: "(716)-516-4621",
  cell: "(887)-227-1124",
  SSN: "566-33-3230",
  picture: "http://api.randomuser.me/portraits/women/78.jpg"
};

it(`renders person`, () => {
  const person = renderer.create(<Person {...user} />).toJSON();
  expect(person).toMatchSnapshot();
});
