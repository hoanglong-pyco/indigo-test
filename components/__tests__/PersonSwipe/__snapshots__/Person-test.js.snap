// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders person 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
        "borderColor": "#ddd",
        "borderRadius": 8,
        "borderWidth": 1,
        "height": 300,
        "width": 280,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#f9f9f9",
          "borderBottomWidth": 1,
          "borderColor": "#ddd",
          "borderTopLeftRadius": 7,
          "borderTopRightRadius": 7,
          "height": 105,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 400,
          "width": 280,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#fff",
            "borderColor": "#ddd",
            "borderRadius": 65,
            "borderWidth": 1,
            "height": 130,
            "justifyContent": "center",
            "marginBottom": 15,
            "marginTop": 15,
            "width": 130,
          },
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://picsum.photos/id/119/240/240",
          }
        }
        style={
          Array [
            Object {
              "backgroundColor": "#fafafa",
              "borderRadius": 60,
              "height": 120,
              "width": 120,
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "height": 110,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 90,
            },
          ]
        }
      >
        <View>
          <Text
            style={
              Array [
                Object {
                  "color": "#9b9b9b",
                  "fontSize": 14,
                  "textAlign": "center",
                },
              ]
            }
          >
            My address is
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#2b2e31",
                  "fontSize": 17,
                  "paddingLeft": 10,
                  "paddingRight": 10,
                  "textAlign": "center",
                },
              ]
            }
          >
            7403 long rapids rd, cedar rapids
          </Text>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "height": 40,
            },
          ]
        }
      >
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <Text />
        </View>
        <View
          accessible={true}
          isTVSelectable={true}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
              "padding": 5,
            }
          }
        >
          <Text />
        </View>
      </View>
    </View>
  </View>
</View>
`;
